Add authentication tests

paolochiodi opened the pull request:
I'm spending quite a long time on these tests, and I'd like to get some early feedback on: 

1- do we even need this kind of tests?
2- is the style ok?

### About the need for these tests
The main reason we started looking into authentication tests was to make sure we are actually enforcing permissions and policies check on our endpoints.
However, the more I write them the more I have two doubts that seems smells of bad tests to me:

a- we are basically testing a config (did we place the correct action in the auth config?)
b- the special cases would be better suited as unit tests for the auth middleware (i.e. the teams case) or for aim-js (wildcards)

I still find some value in these tests because:
a- I don't think we are actually testing a config as per point `a`, because there could be multiple configurations that lead to the correct results and we just need to check if everything works correctly (in an extreme case, a part of the authentication could be moved inside the action instead of the auth middleware and the tests would still work)
b- we are using special cases but we are not actually testing them: we are checking that the chosen authorisation supports all the cases we need

Let me know your thoughts

simoneb commented on service/test/endToEnd/authorization/usersTest.js:
```diff
@@ -0,0 +1,1929 @@
+      }
+
+      lab.before((done) => {
+        userOps.createUser({ name: 'caller', organizationId }, (err, caller) => {
```
if you make these ops functions return promises you can use async/await and shrink the amount of code you need to write here to a great extent

mihaidma commented on service/test/endToEnd/authorization/usersTest.js:
```diff
@@ -0,0 +1,1929 @@
+      })
+
+      lab.after((done) => {
+        userOps.deleteUser({ id: callerId, organizationId }, (err) => {
```
I assume after we refactor tests we'll use a tool to clean all db removing the manual delete. Like: [node-database-cleaner](https://github.com/emerleite/node-database-cleaner)

One issue that I see here is the cleanup running at the end. It could generate junk data. Cleaning should be done before.

mihaidma replied:
We should watch if all the user team creation code can be extracted to a helper.

paolochiodi replied:
lab.after should be always called even after test failures or timeouts - garbage should be generated only in case of major error or tests interruption.

On the helper: I agree. I'll try to find a good approach it

p16 commented:
Hi @paolochiodi 

> I don't think we are actually testing a config as per point `a` [...]

I agree these tests are needed, but having a 2000 lines test file may quickly become unreadable.

One way to overcome this issue and keeping all the tests:

- create one set of fixtures only for these tests that will be loaded in the `before` function
- create an helper function that will make explicit: who is calling and what he is calling (we may not need this)

That said, I think we may also want to reduce the number of tests by removing those that can be part of a [unit test](https://github.com/nearform/labs-authorization/pull/278/files#diff-5568ac6b6a3a9a1d21e6761c4cc51361R327) or those that "duplicates" a behaviour ([here](https://github.com/nearform/labs-authorization/pull/278/files#diff-5568ac6b6a3a9a1d21e6761c4cc51361R104) VS [here](https://github.com/nearform/labs-authorization/pull/278/files#diff-5568ac6b6a3a9a1d21e6761c4cc51361R152))

My idea would be to have tests covering all our endpoints, without testing all possible combination of actions/resources.

paolochiodi commented:
I, too, don't like how long and verbose these tests are, and like @nherment's idea of using some kind of framework or domani language to extract the duplicated code and simplify the code.
Same goes for @mihaidma idea of simplify the creation of the initial content. I would like to use some kind of framework here too, but this is more complex because cases are more diverse and nuanced.

I'll timebox some time on this before moving further ahead with other endpoints

paolochiodi commented:
I incorporated the factory and new test framework and update the PR. I've also removed the `WIP` and  `do not merge` labels as I'm interested in merging this early in order to make the Factory and new test framework available.

In the meantime I'll continue to add tests for other endpoints

p16 commented:
@paolochiodi could you squash the commits? thanks!
